### ISSUE TITLE
Replacing values in Artifact Manager

### ISSUE THREAD

**Issue by jw3** [PDAL/PDAL#1994]
Currently it is not possible to `put` to a key more than once.

Have a patch, if this makes sense Ill prepare a PR

https://github.com/jw3/PDAL/commit/6a502ebbe63af558696c84a98d3f0672ad3f0935

**Comment by hobu** [comment]
This looks ok to me, and you're the only one using it AFAIK, so please make a PR and we'll include it in 1.72 next week.

**Comment by abellgithub** [comment]
I'm thinking that since the point of this is type-safety, it makes sense to type-check what you're replacing:

```
    template <typename T>
    bool replace(const std::string& name, const std::shared_ptr<T> artifact)
    {
        auto it = m_storage.find(name);
        if (it == m_storage.end())
            return false;

        if (!std::dynamic_pointer_cast<T>(it->second))
            return false;
        it->second = artifact;
        return true;
    }
```

**Comment by jw3** [comment]
> it makes sense to type-check what you're replacing

Yeah it does.  

**tl;dr;** How about a `put_or_update` type method that type checks, but doesnt require existing value?


After filing this ticket I was thinking through some of the semantics for the artifact manager api.. thinking more out loud here

I wasn't thrilled about changing the `put` behavior, that was more for a quick fix for our testing here... (1) put was already published as a function that does not replace values, so my change modified the api behavior (2) my solution is not the _right_ answer, it doesnt fix any of the semantic gaps, just patches over an expectation in our code.

The idea of type checking replacements sounds right, but when considering a filter that does not know what came prior to it, a little more might be needed for convenient and clean usage of the api.

If a `put` doesn't replace then it would be nice to have a "put_or_update" method, so that filters dont have to have presence checking logic for every put.

If we have a `put_or_update` that checks type, then perhaps a `replace` method that wouldn't check type, it would just brute force replace a value something with something else.

A `remove` method may be needed as well, especially so if a "put_or_update" is not implemented.

**Comment by jw3** [comment]
So maybe this issue becomes what is the minimum map-like api that the Artifact Manager should expose..

My answer would be 
- get
- put
- put_or_update
- remove

With get and put remaining the same as originally published.

**Comment by abellgithub** [comment]
See: https://github.com/PDAL/PDAL/tree/issue-1994

I can do:

```
    template<typename T>
    bool replaceOrPut(const std::string& name, std::shared_ptr<T> art)
    {
        if (!replace(name, art))
            return put(name, art);
        return true;
    }
```

if that seems helpful, but it's so simple, I'm not sure it's of much value. Up to you.
